sw/device/lib/dif/dif_pinmux.h

```diff
@@ -0,0 +1,502 @@
+   * via a multiplexer matrix inside the Pin Multiplexer.
+   */
+  kDifPinmuxPadringPinKindMio = 0,
+  /**
+   * Dedicated Input/Output pin. Connected directly to a peripheral input
+   * or output, bypassing the multiplexer matrix.
+   */
+  kDifPinmuxPadringPinKindDio,
+} dif_pinmux_padring_pin_kind_t;
+
+/**
+ * Pinmux functionality locking.
+ *
+ * Represents a type of functionality that can be locked inside the
+ * Pin Multiplexer for a given pin/peripheral input.
+ */
+typedef enum dif_pinmux_lock_type {
+  /**
+   * Lock Padring MIO input select for a given peripheral input. This means that
+   * the peripheral input is connection can no longer be changed.
+   */
+  kDifPinmuxLockTypeInsel = 0,
+  /**
+   * Lock peripheral output select for a given Padring MIO output. This means
+   * that the Padring MIO output connection can no longer be changed.
+   */
+  kDifPinmuxLockTypeOutsel,
+  /**
+   * Lock the sleep mode configuration for a given Padring MIO pin.
+   */
+  kDifPinmuxLockTypeMioSleep,
+  /**
+   * Lock the sleep mode configuration for a given Padring DIO pin.
+   */
+  kDifPinmuxLockTypeDioSleep,
+  /**
+   * Lock physical characteristics configuration for a given Padring MIO pin.
+   */
+  kDifPinmuxLockTypeMioPinAttr,
+  /**
+   * Lock physical characteristics configuration for a given Padring DIO pin.
+   */
+  kDifPinmuxLockTypeDioPinAttr,
+} dif_pinmux_lock_type_t;
+
+/**
+ * Pinmux generic index.
+ *
+ * The meaning of this index changes between APIs, and must be documented in
+ * the corresponding function description.
+ */
+typedef uint32_t dif_pinmux_index_t;
+
+/**
+ * Multiplexed Padring input connection select for a peripheral input.
+ */
+typedef uint32_t dif_pinmux_padring_insel_t;
+
+/**
+ * Multiplexed peripheral output connection select for a Padring output.
+ */
+typedef uint32_t dif_pinmux_peripheral_outsel_t;
+
+/**
+ * Pinmux Padring pin attributes.
+ *
+ * Some attributes have defined meanings, others are implementation defined. Not
+ * all pins will support all attributes. The Pinmux DIF does not check for
+ * unsupported attributes. To check to see if an attribute is supported after
+ * enabling it (via `dif_pinmux_padring_attr_enable` or
+ * `dif_pinmux_padring_attr_write`) read back the attribute field (via
+ * `dif_pinmux_padring_attr_get`). If the attribute isn't supported the
+ * attribute will still be disabled (relevant bit unset in the attribute field).
+ */
+typedef enum dif_pinmux_padring_pin_attr {
+  kDifPinmuxPadringPinAttrNone = 0,
+  kDifPinmuxPadringPinAttrIOInvert = 1 << 0,
+  kDifPinmuxPadringPinAttrOpenDrain = 1 << 1,
+  kDifPinmuxPadringPinAttrPullDown = 1 << 2,
+  kDifPinmuxPadringPinAttrPullUp = 1 << 3,
+  kDifPinmuxPadringPinAttrKeeper = 1 << 4,
+  kDifPinmuxPadringPinAttrWeakestDrive = 1 << 5,
+  kDifPinmuxPadringPinAttrWeakDrive = 1 << 6,
+  kDifPinmuxPadringPinAttrStrongDrive = 1 << 7,
+  kDifPinmuxPadringPinAttrStrongestDrive = 1 << 8,
```

Comment:
@msfschaffner does this make sense?

Reply:
This looks good overall, however some of the enums need to be updated.

I quickly crosschecked this with the current RTL implementation / docs, and this is the current bit mapping:
```
0         | Input/output inversion                      | RW
1         | Virtual open drain enable                   | RW
2         | Pull enable                                 | WARL
3         | Pull select (0: down, 1: up)                | WARL
4         | Keeper enable                               | WARL
5         | Schmitt trigger enable                      | WARL
6         | Slew rate (0: slow, 1: fast)                | WARL
8:7       | Drive strengh (00: weakest, 11: strongest)  | WARL
9         | Reserved                                    | WARL
```

Reply:
@msfschaffner do I understand correctly that you mean that the DIF is missing `enable` enum invariants? Or did you mean something else?

I think  the idea is that by picking `kDifPinmuxPadringPinAttrPullDown` it also sets the `Pull enable`, so the DIF enums are slightly more high level, and don't need to match the register bits exactly.

Reply:
The bit mapping doesn't need to match the enum invariants either, as these are mapped into relevant bit offsets in the coming implementation.

Reply:
Ah ok got it. I thought these enums directly correspond to the bit indices in the CSR. 
In that case this looks good to me.

One heads up on the `Reserved` bit: we will likely use this to support an additional pad operating mode on some pads.
Hence, we will have to introduce another enum for that in the future.

Reply:
I think there's been some wavering back and forth over time about the level of abstraction (if any) DIFs should offer over the interface provided by the hardware. It does feel like more closely matching the hardware enums might make life easier though? No need to code up logic to check for illogical / unsupported combinations for instance.

Reply:
There are pros and cons both ways - it seems prudent to not expose register definitions to the header. In this particular case it seems reasonable. I am not sure which way is better but probably would sway more towards the current approach.

Reply:
Hmmm, I think I will match 1:1 CSR bit positions in this case.

Reply:
I went full loop on this issue, and really I think the easiest and better approach is not to match the DIF pad attributes to CSR 1:1. I am committing to it now, and don't think there is much value in discussing this further.

Reply:
@msfschaffner

> One heads up on the Reserved bit: we will likely use this to support an additional pad operating mode on some pads.

Sounds good, please ping me when this needs adding.

Reply:
Let's return to this when reviewing the implementation.